### PR TITLE
enable work queue filter #274

### PR DESCRIPTION
**Pull request by Mikearaya**
in adding this feature I had to create a new endpoint that returns the work types registered in the system at the moment, as I couldn't find a better way to achieve this. if you have any suggestions let me know. the other thing is why not filter `status` as I did with `type` instead of using the reduce function which makes it a little complex at first glance https://github.com/unchainedshop/unchained/blob/master/packages/core-worker/director.js#L165.

also `new` status returns both `Allocated` & `New` works and i don't know if that is intended or not.

### CODE REVIEW

**Comment by Mikearaya** [comment]
@pozylon since you now can filter the work queue i think only one data table is enough instead of separating them into two 2 (new, recent)

File: packages/core-worker/director.js [hunk 1]
```diff
@@ -168,14 +173,36 @@ class WorkerDirector {
...
+        { $skip: skip },
+        {
+          $project: {
+            _id: '$_id',
+            type: '$_id',
+          },
+        },
+      ])
+      .toArray();
+    return typeList;
```
Comment (pozylon) [inline packages/core-worker/director.js:202]: 🤟

**Comment by pozylon** [comment]
Great! I just have one minor thing to ask:

Is there a special reason you have used ``selectTypes`` instead of just ``types`` for the query parameter name?

**Review by pozylon — APPROVED** [review]

**Review by pozylon — CHANGES_REQUESTED** [review]
Oops I was too fast accepting, so the UI still has some issues like lagging when I select/deselect a checkbox and there are multiple warnings thrown when I open browser console. Make sure these are gone. For example sometimes it's not possible to select the types: I get a unique key warning in react:

```
Warning: Each child in a list should have a unique "key" prop.

Check the render method of `WorkList`. See https://reactjs.org/link/warning-keys for more information.
```

Then another request:

The schema is not good yet, please:
1. Remove RegisteredWorkTypes type, use already existing "WorkType" enum which is dynamically extended with all possible types. Also it does not make sense to return a type which has an id and a type with the same value
2. selectTypes param is taking [String] which would mean that these params are accepted: null, undefined, [null, undefined], ["SOMETHING"]. Please use [WorkType!] so it's either null or a list of existing types.

**Comment by Mikearaya** [comment]
:) i'll fix that it must have sliped.
the reason i used `selectedTypes` instead of `types` is because  ` static buildQueueSelector({ status = [], workId, ...rest }) ` is used by other part of the codebase that pass it the actually type. plus it's more descriptive :)

**Comment by pozylon** [comment]
Force pushed this, make sure you hard reset when you continue working on this

**Comment by Mikearaya** [comment]
@pozylon i'm almost done with this PR too except bug in infinite-scroll, it keeps on repeating the same row when updating the table (checking and unchecking) status boxes. glance over it if you have the energy please :pleading_face:

**Comment by pozylon** [comment]
Okay so I think it's not quite done and if you change some UI elements maybe that issue with the infinite scroll disappears:

1. You have decided to re-use the SearchDropdown for the types. I think that the SearchDropdown is only really required if you have a fulltext search that needs to emit API calls. I'd go for native semantic UI component because you don't need to refetch workTypes and it's more like a selection instead of a dropdown. So consider refactoring this that workTypes is fetched just once when the UI pops up. See https://codesandbox.io/s/etd1x?module=/example.js
2. Maybe from a UI perspective it makes more sense to see the selectTypes as filters that you can set, so you could also use the default Dropdown with the same multi selection feature like in https://codesandbox.io/s/etd1x?module=/example.js. If you select one or more than it is sent along the query as selectTypes and if no type is selected it will show all types.

**Comment by pozylon** [comment]
Then please have a look at my comment regarding the RegisteredWorkType type, which i think is obsolete:

> Then another request:

You could improve the type stuff like this:

Remove RegisteredWorkTypes type, use already existing "WorkType" enum which is dynamically extended with all possible types. Also it does not make sense to return a type which has an id and a type with the same value
selectTypes param is taking [String] which would mean that these params are accepted: null, undefined, [null, undefined], ["SOMETHING"]. Please use [WorkType!] so it's either null or a list of existing types.

**Comment by Mikearaya** [comment]
Thanks, i've made all the necessary adjustments now, and also I have decided to name the resolver `activeWorkTypes` instead of `workTypes` because it's more descriptive.

File: packages/api/schema/query.js [hunk 2]
```diff
@@ -303,6 +304,10 @@ export default [
       Get a specific work unit by ID
       """
       work(workId: ID!): Work
+      """
+      Get List of currently registered worker plugins
+      """
+      activeWorkTypes: [WorkType]!
```
Comment (pozylon) [inline packages/api/schema/query.js:310]: [WorkType!]!

Comment (Mikearaya) [reply packages/api/schema/query.js:310]: I made it like that in the case where there is no worktype so user should expect empty array or array of worktype.

Comment (Mikearaya) [reply packages/api/schema/query.js:310]: I know it unlikely there will not be any work but its the right logic.

**Review by pozylon — CHANGES_REQUESTED** [review]
When I add a new job the list does not update correctly, maybe something about the keys?

```
mutation {
  addWork(type: HEARTBEAT) {
    _id
  }
}
```

File: packages/api/schema/query.js (hunk 2 above)
Comment (pozylon) [reply packages/api/schema/query.js:310]: [WorkType]! = [null], I don't think you want to return an array of null

**Comment by Mikearaya** [comment]
@pozylon if you are referring to the work type filter dropdown, did you refresh the page? because that list is filled on the initial render. other than that it's working as expected. can you check again?

**Comment by Mikearaya** [comment]
added 10 seconds of poll interval for work type dropdown so no need of a refresh.

**Comment by pozylon** [comment]
Nood good yet, still same issue, it's not that it's not polling, it was polling before but there are multiple issues in the console, react errors and such and I think there is _id conflicts and load more is broken.

**Comment by pozylon** [comment]
If you add a heartbeat work for example, a new line appears but the lines are rendered wrong (new timestamp not showing).

**Comment by Mikearaya** [comment]
@pozylon can you check out the current state of this PR and let me know if there is anything it's missing. The problem existed before it's just that we haven't noticed it because there were separate tables for each.

I have made work queue query to never stop polling because it needs to update the moment new work is add and that not how infinite scroll works. It requires for load more to be called for it to update.

**Comment by pozylon** [comment]
Maybe the infinite table just does not work well together with polling and the best thing is getting rid of the infinite table

**Review by pozylon — APPROVED** [review]
I think we can go with this for the moment, let's see how it behaves in prod and if we have probs we just remove the infinite loading and replace it with some kind of paging

**Comment by Mikearaya** [comment]
Sounds good to me.

**Pull request merged**